File: src/aizk/extract/journal.py
import re
from datetime import UTC, datetime

from . import ontology
from .models import ExtractedEntity, TimedFact

# Dated entries may include a parenthesized label after the date.
_JOURNAL_LINE = re.compile(r"^-\s*(\d{4}-\d{2}-\d{2})(?:\s*\([^)]*\))?:\s*(.+)$", re.MULTILINE)

_AREA_TAG = re.compile(r"(?<!\w)#area(?!\w)", re.IGNORECASE)
_PROJECT_TAG = re.compile(r"(?<!\w)#project(?!\w)", re.IGNORECASE)


def has_journal_entries(text: str) -> bool:
    """Whether text contains at least one dated journal entry."""
    return _JOURNAL_LINE.search(text) is not None


def declared_type(text: str) -> str | None:
    """The structural type a note's own tags declare, Area or Project, else None for an
    ordinary note the extractor is left to characterize."""
    if _AREA_TAG.search(text):
        return ontology.AREA
    if _PROJECT_TAG.search(text):
        return ontology.PROJECT
    return None


def title_entity(title: str, declared: str | None) -> ExtractedEntity:
    """The note's own title as an entity, typed by the structural tag it declares or Concept."""
    return ExtractedEntity(name=title, type=declared or ontology.CONCEPT)


def journal_facts(chunk_text: str, title: str) -> list[TimedFact]:
    """Parse a chunk's dated journal lines into facts logged against the note's title entity."""
    facts = []
    for date_text, statement in _JOURNAL_LINE.findall(chunk_text):
        valid_from = datetime.strptime(date_text, "%Y-%m-%d").replace(tzinfo=UTC)
        facts.append(
            TimedFact(
                subject=title,
                predicate=ontology.OBSERVES,
                object="",
                statement=statement.strip(),
                valid_from=valid_from,
                valid_to=None,
            )
        )
    return facts

File: archive/ts/src/lib/server/extract/journal.ts
// Journal parsing ported from the Python extract/journal.py: dated bullet entries become
// observation facts against the note's own title entity, and #area/#project tags declare
// the note's structural type.
import { AREA, CONCEPT, OBSERVES, PROJECT } from './ontology';
import type { ExtractedEntity, TimedFact } from './models';

// Dated entries may include a parenthesized label after the date.
const JOURNAL_LINE = /^-\s*(\d{4}-\d{2}-\d{2})(?:\s*\([^)]*\))?:\s*(.+)$/gm;

const AREA_TAG = /(?<!\w)#area(?!\w)/i;
const PROJECT_TAG = /(?<!\w)#project(?!\w)/i;

// Whether text contains at least one dated journal entry. The shared pattern is global for
// matchAll, so the probe resets its cursor on both sides of the test.
export const hasJournalEntries = (text: string): boolean => {
	JOURNAL_LINE.lastIndex = 0;
	const found = JOURNAL_LINE.test(text);
	JOURNAL_LINE.lastIndex = 0;
	return found;
};

// The structural type a note's own tags declare, Area or Project, else null for an
// ordinary note the extractor is left to characterize.
export const declaredType = (text: string): string | null => {
	if (AREA_TAG.test(text)) return AREA;
	if (PROJECT_TAG.test(text)) return PROJECT;
	return null;
};

// The note's own title as an entity, typed by the structural tag it declares or Concept.
export const titleEntity = (title: string, declared: string | null): ExtractedEntity => ({
	name: title,
	type: declared ?? CONCEPT,
	suggestedType: null
});

// Parse a chunk's dated journal lines into facts logged against the note's title entity.
export const journalFacts = (chunkText: string, title: string): TimedFact[] => {
	JOURNAL_LINE.lastIndex = 0;
	return [...chunkText.matchAll(JOURNAL_LINE)].map(([, dateText, statement]) => {
		const [year, month, day] = dateText.split('-').map(Number);
		return {
			subject: title,
			predicate: OBSERVES,
			object: '',
			statement: statement.trim(),
			quote: null,
			validFrom: new Date(Date.UTC(year, month - 1, day)),
			validTo: null,
			kind: 'world' as const
		};
	});
};
